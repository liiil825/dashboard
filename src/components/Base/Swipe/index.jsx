import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import _ from 'lodash';

import styles from './index.scss';

export default class Swipe extends React.Component {
  static propTypes = {
    autoPlay: PropTypes.bool,
    height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    hoverStop: PropTypes.bool,
    showDots: PropTypes.bool,
    speed: PropTypes.number,
    unit: PropTypes.string,
    width: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
  };

  static defaultProps = {
    autoPlay: true,
    hoverStop: true,
    showDots: true,
    speed: 5000,
    width: '100vw',
    height: '100vh',
    unit: 'px'
  };

  state = {
    hasTransition: false,
    current: 1
  };

  componentDidMount() {
    if (this.props.autoPlay) {
      this.play();
    }
  }

  componentWillUnmount() {
    this.pause();
  }

  get hasChildren() {
    return this.props.children && this.props.children.length > 0;
  }

  get total() {
    return this.props.children.length;
  }

  get widthNumber() {
    const { width } = this.props;
    if (_.isNumber(width)) {
      return width;
    }
    return parseFloat(width);
  }

  get widthUnit() {
    const { width, unit } = this.props;
    if (_.isNumber(width)) {
      return unit;
    }

    return _.last(width.split(/\d./));
  }

  get width() {
    return `${this.widthNumber}${this.widthUnit}`;
  }

  get height() {
    const { height, unit } = this.props;
    if (_.isNumber(height)) {
      return `${height}${unit}`;
    }
    return height;
  }

  get totalWidth() {
    const total = this.total + 2;
    return `${this.widthNumber * total}${this.widthUnit}`;
  }

  get transform() {
    const { current } = this.state;
    const translateX = `${current * -1 * this.widthNumber}${this.widthUnit}`;
    return `translateX(${translateX})`;
  }

  changeIndex = (plusNum, idx = null) => {
    let { current } = this.state;
    const { total } = this;
    let next = null;
    if (_.isNumber(plusNum)) {
      current += plusNum;
    } else if (_.isNumber(idx)) {
      current = idx;
    }
    if (current >= total + 1) {
      current = total + 1;
      next = 1;
    } else if (current <= 0) {
      current = 0;
      next = total;
    }
    this.setState({
      hasTransition: true,
      current
    });
    if (next) {
      setTimeout(() => {
        this.setState({
          hasTransition: false,
          current: next
        });
      }, 500);
    }
  };

  play = () => {
    this.interval = setInterval(() => this.changeIndex(1), this.props.speed);
  };

  pause = () => {
    clearInterval(this.interval);
  };

  renderDots() {
    const { showDots, dotsOuter } = this.props;
    if (!showDots) return null;

    return (
      <div className={classnames(styles.dotsOuter, dotsOuter)}>
        {_.range(this.total).map(index => (
          <span
            key={index}
            onClick={() => this.changeIndex(null, index + 1)}
            className={classnames(
              styles.dot,
              {
                [styles.active]: this.state.current - 1 === index
              },
              this.props.dotCls
            )}
          />
        ))}
      </div>
    );
  }

  renderContent() {
    const { children, contentCls } = this.props;
    const appendedChildren = [
      children[children.length - 1],
      ...children,
      children[0]
    ];

    return React.Children.map(appendedChildren, (child, idx) => {
      const childClone = React.cloneElement(child);

      return (
        <div
          className={classnames(styles.content, contentCls)}
          style={{ width: this.width }}
          key={`swipe-${child.key}-${idx}`}
        >
          {childClone}
        </div>
      );
    });
  }

  render() {
    const {
      className, autoPlay, hoverStop, contentOuter
    } = this.props;
    const { hasTransition } = this.state;

    if (!this.hasChildren) {
      return null;
    }
    return (
      <div
        className={classnames(styles.container, className)}
        onMouseEnter={autoPlay && hoverStop ? this.pause : _.noop}
        onMouseLeave={autoPlay && hoverStop ? this.play : _.noop}
        style={{ height: this.height }}
      >
        <div
          className={classnames(
            styles.contentOuter,
            {
              [styles.contentTransition]: hasTransition
            },
            contentOuter
          )}
          style={{
            width: this.totalWidth,
            transform: this.transform
          }}
        >
          {this.renderContent()}
        </div>
        {this.renderDots()}
      </div>
    );
  }
}
