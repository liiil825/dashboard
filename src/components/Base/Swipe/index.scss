@import '~scss/vars';

.container {
  position: relative;
  overflow: hidden;
}
.contentOuter {
  position: absolute;
}
.content {
  display: inline-flex;
}

.contentTransition {
  transition: all 0.5s ease-in-out;
}
.dotsOuter {
  position: absolute;
  bottom: 3vh;
  display: flex;
  justify-content: center;
  width: 100%;
  .dot + .dot {
    margin-left: 20px;
  }
}
.dot {
  width: 40px;
  height: 20px;
  background: linear-gradient(
    transparent 0%,
    transparent 40%,
    $Y75 40%,
    $Y75 60%,
    transparent 60%,
    transparent
  );
  opacity: 0.12;
  cursor: pointer;
  &.active {
    opacity: 1;
  }
}
