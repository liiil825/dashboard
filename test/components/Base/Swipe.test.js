import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import Swipe from 'components/Base/Swipe';

describe('Base/Swipe', () => {
  it('basic render', () => {
    const wrapper = render(
      <Swipe>
        <div>1</div>
        <div>2</div>
        <div>3</div>
        <div>4</div>
      </Swipe>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('no render', () => {
    const wrapper = render(<Swipe />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('render no dots', () => {
    const wrapper = render(
      <Swipe showDots={false} height={500} width={500} unit="px">
        <div>1</div>
        <div>2</div>
        <div>3</div>
        <div>4</div>
      </Swipe>
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('call click dot', () => {
    const wrapper = mount(
      <Swipe>
        <div>1</div>
        <div>2</div>
        <div>3</div>
        <div>4</div>
      </Swipe>
    );
    const dotsComponent = wrapper.find('.dotsOuter');
    dotsComponent.childAt(3).simulate('click');
    expect(wrapper.state().current).toBe(4);

    dotsComponent.childAt(0).simulate('click');
    expect(wrapper.state().current).toBe(1);
    wrapper.unmount();
  });
});
