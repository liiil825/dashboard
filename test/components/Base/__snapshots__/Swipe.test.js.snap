// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base/Swipe basic render 1`] = `
<div
  class="container"
  style="height:100vh"
>
  <div
    class="contentOuter"
    style="width:600w;transform:translateX(-100w)"
  >
    <div
      class="content"
      style="width:100w"
    >
      <div>
        4
      </div>
    </div>
    <div
      class="content"
      style="width:100w"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="content"
      style="width:100w"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="content"
      style="width:100w"
    >
      <div>
        3
      </div>
    </div>
    <div
      class="content"
      style="width:100w"
    >
      <div>
        4
      </div>
    </div>
    <div
      class="content"
      style="width:100w"
    >
      <div>
        1
      </div>
    </div>
  </div>
  <div
    class="dotsOuter"
  >
    <span
      class="dot active"
    />
    <span
      class="dot"
    />
    <span
      class="dot"
    />
    <span
      class="dot"
    />
  </div>
</div>
`;

exports[`Base/Swipe no render 1`] = `null`;

exports[`Base/Swipe render no dots 1`] = `
<div
  class="container"
  style="height:500px"
>
  <div
    class="contentOuter"
    style="width:3000px;transform:translateX(-500px)"
  >
    <div
      class="content"
      style="width:500px"
    >
      <div>
        4
      </div>
    </div>
    <div
      class="content"
      style="width:500px"
    >
      <div>
        1
      </div>
    </div>
    <div
      class="content"
      style="width:500px"
    >
      <div>
        2
      </div>
    </div>
    <div
      class="content"
      style="width:500px"
    >
      <div>
        3
      </div>
    </div>
    <div
      class="content"
      style="width:500px"
    >
      <div>
        4
      </div>
    </div>
    <div
      class="content"
      style="width:500px"
    >
      <div>
        1
      </div>
    </div>
  </div>
</div>
`;
